airbyte-integrations/connectors/source-mixpanel/source_mixpanel/source.py

```diff
@@ -0,0 +1,839 @@
+        }
+
+
+class Funnels(DateSlicesMixin, IncrementalMixpanelStream):
+    """List the funnels for a given date range.
+    API Docs: https://developer.mixpanel.com/reference/funnels#funnels-query
+    Endpoint: https://mixpanel.com/api/2.0/funnels
+    """
+
+    primary_key = ["funnel_id", "date"]
+    data_field = "data"
+    cursor_field = "date"
+    min_date = "90"  # days
+
+    def path(self, **kwargs) -> str:
+        return "funnels"
+
+    def funnel_slices(self, sync_mode) -> List[dict]:
+        funnel_slices = FunnelsList(authenticator=self.authenticator).read_records(sync_mode=sync_mode)
+        funnel_slices = list(funnel_slices)  # [{'funnel_id': <funnel_id1>, 'name': <name1>}, {...}]
+
+        # save all funnels in dict(<funnel_id1>:<name1>, ...)
+        self.funnels = dict((funnel["funnel_id"], funnel["name"]) for funnel in funnel_slices)
+
+        return funnel_slices
+
+    def stream_slices(
+        self, sync_mode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Mapping[str, Any]]]]:
+        """Return stream slices which is a combination of all funnel_ids and related date ranges, like:
+        stream_slices = [
+            {   'funnel_id': funnel_id1_int,
+                'funnel_name': 'funnel_name1',
+                'start_date': 'start_date_1'
+                'end_date': 'end_date_1'
+            },
+            {   'funnel_id': 'funnel_id1_int',
+                'funnel_name': 'funnel_name1',
+                'start_date': 'start_date_2'
+                'end_date': 'end_date_2'
+            }
+            ...
+            {   'funnel_id': 'funnel_idX_int',
+                'funnel_name': 'funnel_nameX',
+                'start_date': 'start_date_1'
+                'end_date': 'end_date_1'
+            }
+            ...
+        ]
+
+        # NOTE: funnel_id type:
+        #    - int in funnel_slice
+        #    - str in stream_state
+        """
+        stream_state = stream_state or {}
+
+        # One stream slice is a combination of all funnel_slices and date_slices
+        stream_slices = []
+        funnel_slices = self.funnel_slices(sync_mode)
+        # date_slices = self.date_slices(sync_mode, cursor_field=cursor_field, stream_state=stream_state)
+        for funnel_slice in funnel_slices:
+            # get single funnel state
+            funnel_id = str(funnel_slice["funnel_id"])
+            funnel_state = stream_state.get(funnel_id)
+            date_slices = super().stream_slices(sync_mode, cursor_field=cursor_field, stream_state=funnel_state)
+            for date_slice in date_slices:
+                stream_slices.append({**funnel_slice, **date_slice})
+
+        # reset rate_limit if we expect less requests (1 req per stream) than it is allowed by API rate_limit
+        if len(stream_slices) < self.rate_limit:
+            self.rate_limit = 3600  # queries per hour (1 query per sec)
```

Comment:
where does 3600 come from? why set this if `len(stream_slices) < self.rate_limit`?

Reply:
Planned API limits:
Export stream - 60 reqs/hour (1 req for 60 sec).
Another streams - 400 reqs/hour (1 req for 9 sec)

When I used our default exponential backoff approach I got 'too many requests' error a few times.

So that's why I decided to run requests with:

- 1 sec delay if we know that the total number of requests is going to be less than API rate limit.
- 'planned' sec - delay if there are too many planned requests and we need to overcome the error.


I am still new in connector development so I really need your advice here.

Could you please confirm that it is better to remove this suggested approach and just use the default exponential backoff approach

Thank you!